### PR TITLE
Remove swap after dup.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by ekpyron** [comment]
I guess this one makes sense, but it's not the one needed for https://github.com/ethereum/solidity/pull/9378, is it?

**Comment by chriseth** [comment]
I think it is - there is a second rule that removes pop after dup.

**Comment by ekpyron** [comment]
Ok, right! Tests are failing though...

File: test/cmdlineTests/dup_opt_peephole/output [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+Text representation:
+    /* "yul_stack_opt_peephole/input.sol":25:26   */
+  0x00
+    /* "yul_stack_opt_peephole/input.sol":12:27   */
+  calldataload
+    /* "yul_stack_opt_peephole/input.sol":34:35   */
+  dup1
+    /* "yul_stack_opt_peephole/input.sol":29:35   */
+  swap1
```
Comment (axic) [inline test/cmdlineTests/dup_opt_peephole/output]: This still has dup1+swap1?

File: test/cmdlineTests/dup_opt_peephole/output [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  pop
```
Comment (axic) [inline test/cmdlineTests/dup_opt_peephole/output:15]: Also dup1+swap1+pop doesn't seem to make all that much sense.

File: libevmasm/PeepholeOptimiser.cpp [hunk 3]
```diff
@@ -205,6 +205,31 @@ struct SwapComparison: SimplePeepholeOptimizerMethod<SwapComparison, 2>
...
+		AssemblyItem const& _dupN,
+		AssemblyItem const& _swapN,
+		std::back_insert_iterator<AssemblyItems> _out
+	)
+	{
+		if (
+			SemanticInformation::isDupInstruction(_dupN) &&
+			SemanticInformation::isSwapInstruction(_swapN) &&
+				uint8_t(_dupN.instruction()) - uint8_t(Instruction::DUP1) ==
+				uint8_t(_swapN.instruction()) - uint8_t(Instruction::SWAP1)
```
Comment (axic) [inline libevmasm/PeepholeOptimiser.cpp]: Nitpick: this is a weird formatting with the extra indentation.

**Comment by chriseth** [comment]
Updated.

File: libevmasm/PeepholeOptimiser.cpp (hunk 3 above)
Comment (ekpyron) [reply libevmasm/PeepholeOptimiser.cpp]: It is correct according to our coding style due to the parenthesis around it... but yeah, it's a bit weird here :-).

This weird subtraction in general might be worth a helper in general, but maybe not, and I'd be fine with keeping it as is for this PR.

**Review by ekpyron — DISMISSED** [review]

File: libevmasm/PeepholeOptimiser.cpp (hunk 3 above)
Comment (axic) [reply libevmasm/PeepholeOptimiser.cpp]: There was no parenthesis the time I left this comment. Github still shows the original code:

Screenshot 2020-07-13 at 13 44 12

**Comment by chriseth** [comment]
Actually there is a helper...

**Review by ekpyron — APPROVED** [review]

**Comment by leonardoalt** [comment]
Fuzzer build error. Does it need a rebase?

**Comment by ekpyron** [comment]
> Fuzzer build error. Does it need a rebase?

It's unrelated to the change here in any case - I think it's in the process of being fixed on develop, but not sure the fix is there yet.

**Comment by leonardoalt** [comment]
https://github.com/ethereum/solidity/pull/9392 was merged

File: Changelog.md [hunk 4]
```diff
@@ -10,6 +10,7 @@ Bugfixes:
 Compiler Features:
  * Build System: Update internal dependency of jsoncpp to 1.9.3.
  * Optimizer: Add rule to remove shifts inside the byte opcode.
+ * Peephole Optimizer: Add rule to remove swap after dup.
```
Comment (axic) [inline Changelog.md:13]: Perhaps the following once we do a changelog review:

```suggestion
 * Peephole Optimizer: Add rule to remove ``SWAPn`` after ``DUPn``.
```

**Pull request merged**